### PR TITLE
feat: php multiparts

### PR DESCRIPTION
**Pull request by byawitz**
## What does this PR do?
Changes `createExecution` to send and parse multipart data 

Code example 👇
```php
<?php

use Appwrite\Client;
use Appwrite\Payload;
use Appwrite\Services\Functions;

$client = (new Client())
    ->setEndpoint('https://cloud.appwrite.io/v1') // Your API Endpoint
    ->setProject('&lt;YOUR_PROJECT_ID&gt;') // Your project ID
    ->setSession(''); // The user session to authenticate with

$functions = new Functions($client);

$result = $functions->createExecution(
    functionId: '<FUNCTION_ID>',
    body: Payload::fromJson([]), // optional
    async: false, // optional
    path: '<PATH>', // optional
    method: ExecutionMethod::GET(), // optional
    headers: '', // optional
    scheduledAt: '' // optional
);
```

## Test Plan

image

### CODE REVIEW

File: mock-server/app/http.php [hunk 1]
```diff
@@ -317,12 +316,11 @@
     ->param('x', '', new Text(100), 'Sample string param')
     ->param('y', '', new Integer(true), 'Sample numeric param')
     ->param('z', null, new ArrayList(new Text(256), APP_LIMIT_ARRAY_PARAMS_SIZE), 'Sample array param')
-    ->param('file', [], new File(), 'Sample file param', skipValidation: true)
+    ->param('payload', [], new File(), 'Sample file param', skipValidation: true)
```
Comment (christyjacob4) [inline mock-server/app/http.php:319]: this change is not required right? we can keep the param name as file itself?

File: tests/languages/php/test.php [hunk 2]
```diff
@@ -118,6 +118,11 @@
 );
 echo $url . "\n";
 
+$response = $general->multipart();
+echo "{$response['x']}\n";
+$hash = md5($response['responseBody']->toBinary());
+echo "{$hash}\n";
+
```
Comment (christyjacob4) [inline tests/languages/php/test.php:125]: cool this is good, but this is an example of how we receive multipart data right? Don't we meed to have a test for  sending multipart as well ?

File: templates/php/base/params.twig [hunk 3]
```diff
@@ -16,6 +20,9 @@
         {%~ for parameter in method.parameters.header %}
         $apiHeaders['{{ parameter.name }}'] = ${{ parameter.name | caseCamel | escapeKeyword }};
         {%~ endfor %}
+        {%~ if method.name | lower == "createexecution" %}
+        $apiHeaders['accept'] = 'multipart/form-data';
+        {%~ endif %}
```
Comment (christyjacob4) [inline templates/php/base/params.twig:25]: I think we should move this code to the client class, inside  the case for multi part 

Screenshot 2024-08-30 at 11 40 04 AM

File: templates/php/src/Payload.php.twig [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+    {
+        $instance = new Payload();
+        $instance->path = null;
+        $instance->data = $data;
+        $instance->mimeType = $mimeType;
+        $instance->filename = $filename;
+        return $instance;
+    }
+
+    public static function fromJson(array $data) {
```
Comment (christyjacob4) [inline templates/php/src/Payload.php.twig:52]: missing return type. I think we can use `:self` right ?

File: templates/php/base/params.twig (hunk 3 above)
Comment (christyjacob4) [reply templates/php/base/params.twig:25]: Or we should pass some sort of an attribute from the specs to make this check easier

File: templates/php/src/Payload.php.twig [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+    {
+        return $this->mimeType;
+    }
+
+    public function getFilename(): ?string
+    {
+        return $this->filename;
+    }
+
+    public static function fromFile(string $path, ?string $mimeType = null, ?string $filename = null): Payload
```
Comment (christyjacob4) [inline templates/php/src/Payload.php.twig:32]:
```suggestion
    public static function fromFile(string $path, ?string $mimeType = null, ?string $filename = null): self
```

File: templates/php/src/Payload.php.twig [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+    {
+        $instance = new Payload();
+        $instance->path = $path;
+        $instance->data = null;
+        $instance->mimeType = $mimeType;
+        $instance->filename = $filename;
+        return $instance;
+    }
+
+    public static function fromBinary(string $data, ?string $mimeType = null, ?string $filename = null): Payload
```
Comment (christyjacob4) [inline templates/php/src/Payload.php.twig:42]:
```suggestion
    public static function fromBinary(string $data, ?string $mimeType = null, ?string $filename = null): self
```

File: templates/php/src/Payload.php.twig [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+        $instance->data = $data;
+        $instance->mimeType = $mimeType;
+        $instance->filename = $filename;
+        return $instance;
+    }
+
+    public static function fromJson(array $data) {
+        $instance = new Payload();
+        $instance->path = null;
+        $instance->data = json_encode($data);
```
Comment (christyjacob4) [inline templates/php/src/Payload.php.twig:55]: lets set the mime type to `application/json`

File: templates/php/src/Payload.php.twig [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+        return $instance;
+    }
+
+    public static function fromString(string $data) {
```
Comment (christyjacob4) [inline templates/php/src/Payload.php.twig:59]: missing return type

File: templates/php/src/Payload.php.twig [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+        $instance = new Payload();
+        $instance->path = null;
+        $instance->data = $data;
+        return $instance;
+    }
+
+    public function toBinary(): string
+    {
+        return $this->data;
```
Comment (christyjacob4) [inline templates/php/src/Payload.php.twig:68]: is binary represented as a string in PHP? is there no other data type like a buffer or uint8 array or something?

File: templates/php/src/Services/Service.php.twig [hunk 10]
```diff
@@ -53,7 +53,7 @@ class {{ service.name | caseUcfirst }} extends Service
         );
 
         {{~ include('php/base/params.twig') -}}
-        {%~ if 'multipart/form-data' in method.consumes %}
+        {%~ if 'multipart/form-data' in method.consumes and method.name | lower != "createexecution" %}
```
Comment (christyjacob4) [inline templates/php/src/Services/Service.php.twig:56]: lets see if there is a better way to do this check

File: mock-server/app/http.php (hunk 1 above)
Comment (loks0n) [reply mock-server/app/http.php:319]: This a quick test to determine that we aren't using hardcoded checks against the `file` param name

File: tests/languages/php/test.php (hunk 2 above)
Comment (loks0n) [reply tests/languages/php/test.php:125]: We already have a test for this, the `upload` function

File: templates/php/src/Client.php.twig [hunk 11]
```diff
@@ -240,4 +246,35 @@ class Client
...
+            $matched = preg_match('/name="?(?<name>\w+)/s', $part, $matches);
+            if ($matched) {
+                array_shift($lines);
+                if(isset($lines[0]) && $lines[0] === 'Content-Type: application/json'){
+                    array_shift($lines);
+                    $headers = json_decode(implode($lines), true);
+                    $headers = array_combine(
+                        array_map(fn($header)=>  $header['name'], $headers),
+                        array_map(fn($header)=>  $header['value'], $headers)
+                    );
```
Comment (abnegate) [inline templates/php/src/Client.php.twig:266]: We should support any JSON, not just objects with name/value

File: templates/php/src/Client.php.twig [hunk 12]
```diff
@@ -240,4 +246,35 @@ class Client
...
+                    $data[$matches['name']] = $headers;
+                    continue;
+                }
+                $data[$matches['name']] = implode("\r\n",$lines) ?? '';;
+            }
+        }
+
+        $data['responseStatusCode'] = (int) ($data['responseStatusCode'] ?? '');
+        $data['duration'] = ((float) ($data['duration'] ?? ''));
+        $data['responseBody'] = Payload::fromString($data['responseBody'] ?? '');
```
Comment (abnegate) [inline templates/php/src/Client.php.twig:276]: We should only set these if they already exist in $data

File: templates/php/src/Payload.php.twig (hunk 9 above)
Comment (byawitz) [reply templates/php/src/Payload.php.twig:68]: It's a string, yes

**Review by christyjacob4 — APPROVED** [review]

**Pull request merged**